subsys/fs/nvs/nvs.c

```diff
@@ -0,0 +1,835 @@
+
+	if (fs->max_len > max_item_len) {
+		SYS_LOG_ERR("Configuration error - max item size");
+		return -EINVAL;
+	}
+	for (i = 0; i < fs->sector_count; i++) {
+		rc = _nvs_sector_hdr_get(fs, i * fs->sector_size, &sector_hdr);
+		if (rc) {
+			return rc;
+		}
+		if (sector_hdr.fd_magic != fs->magic) {
+			continue;
+		}
+		active_sector_cnt++;
+		if (entry_sector < 0) {
+			entry_sector = i;
+			entry_sector_id = sector_hdr.fd_id;
+			active_sector_id = sector_hdr.fd_id;
+			continue;
+		}
+		if (NVS_ID_GT(sector_hdr.fd_id, active_sector_id)) {
+			active_sector_id = sector_hdr.fd_id;
+		}
+		if (NVS_ID_GT(entry_sector_id, sector_hdr.fd_id)) {
+			entry_sector = i;
+			entry_sector_id = sector_hdr.fd_id;
+		}
+	}
+	if (entry_sector < 0) { /* No valid sectors found */
+		SYS_LOG_DBG("No valid sectors found, initializing sectors");
+		for (addr = 0; addr < (fs->sector_count * fs->sector_size);
+			addr += fs->sector_size) {
+			/* first check if used, only erase if it is */
+			if (_nvs_sector_is_used(fs, addr)) {
+				rc = _nvs_flash_erase(fs, addr,
+						      fs->sector_size);
+				if (rc) {
+					return rc;
+				}
+			}
+		}
+		rc = _nvs_sector_init(fs, 0);
+		if (rc) {
+			return rc;
+		}
+		entry_sector = 0;
+		active_sector_id = fs->sector_id;
+	}
+	fs->entry_sector = entry_sector;
+	fs->sector_id = active_sector_id;
+
+	/* Find the first empty entry */
+	nvs_set_start_entry(fs, &entry);
+	entry.id = NVS_ID_EMPTY;
+	rc = nvs_walk_entry(fs, &entry);
+	if (rc) {
+		return rc;
+	}
+	fs->write_location = _nvs_head_addr_in_flash(fs, &entry);
+	if (active_sector_cnt == fs->sector_count) {
+		/* one sector should always be empty, unless power was cut
+		 * during garbage collection start gc again on the last
+		 * sector.
+		 */
+		SYS_LOG_DBG("Restarting garbage collection");
+		addr = fs->entry_sector * fs->sector_size;
+		_nvs_entry_sector_advance(fs);
+		rc = _nvs_gc(fs, addr);
+		if (rc) {
+			return rc;
+		}
+		rc = _nvs_flash_erase(fs, addr, fs->sector_size);
+		if (rc) {
+			return rc;
+		}
+	}
+
+	SYS_LOG_INF("maximum storage length %d bytes", fs->max_len);
+	SYS_LOG_INF("write-align: %d, write-addr: %x",
+		fs->write_block_size, fs->write_location);
+	SYS_LOG_INF("entry sector: %d, entry sector ID: %d",
+		fs->entry_sector, fs->sector_id);
+
+	k_mutex_init(&fs->nvs_lock);
+
+	return 0;
+}
+
+int nvs_append(struct nvs_fs *fs, struct nvs_entry *entry)
+{
+	int rc;
+	u16_t required_len, extended_len, hdr_len, slt_len;
+	struct _nvs_data_hdr data_hdr;
+
+	rc = 0;
+
+	k_mutex_lock(&fs->nvs_lock, K_FOREVER);
+	required_len = _nvs_entry_len_in_flash(fs, entry->len);
+
+	/* the space available should be big enough to fit the data + header
+	 * and slot of next data
+	 */
+	hdr_len = _nvs_len_in_flash(fs, sizeof(struct _nvs_data_hdr));
+	slt_len = _nvs_len_in_flash(fs, sizeof(struct _nvs_data_slt));
+	extended_len = required_len + hdr_len + slt_len;
+
+	if ((fs->sector_size - (fs->write_location & (fs->sector_size - 1))) <
```

Comment:
Hi @Laczen, I have a comment here. I think there is a possibility that the last entry will occupy all remaining space in a sector. In that case, not only it would not be possible to delete anything, but also it would be impossible to GC since we couldn't "close" the sector. If that is the case, what do you think about reserving a few bytes to make sure that nvs can always "close" a sector?
(edit: I mistakenly started a review instead of adding a comment, sorry).